File: src/main/java/org/tindalos/principle/domain/detector/submodulesblueprint/SubmodulesBlueprintViolationDetector.java
package org.tindalos.principle.domain.detector.submodulesblueprint;

import java.util.List;
import java.util.Map;
import java.util.Set;

import org.tindalos.principle.domain.coredetector.CheckInput;
import org.tindalos.principle.domain.coredetector.CheckResult;
import org.tindalos.principle.domain.coredetector.Detector;
import org.tindalos.principle.domain.expectations.DesignQualityExpectations;
import org.tindalos.principle.domain.expectations.SubmodulesBlueprint;

import com.google.common.collect.Maps;
import com.google.common.collect.Sets;

public class SubmodulesBlueprintViolationDetector implements Detector {

	private final SubmodulesFactory submoduleFactory;

	public SubmodulesBlueprintViolationDetector(SubmodulesFactory submoduleFactory) {
		this.submoduleFactory = submoduleFactory;
	}

	public CheckResult analyze(CheckInput checkInput) {

		SubmodulesBlueprint submodulesBlueprint = checkInput.getSubmodulesBlueprint();
		
		List<Submodule> submodules = submoduleFactory.buildSubmodules(submodulesBlueprint.getSubmodulesDefinitionLocation(), checkInput.getPackages(),
				checkInput.getBasePackage());

		Map<Submodule, Set<Submodule>> allViolations = Maps.newHashMap();
		for (Submodule submodule : submodules) {
			Set<Submodule> violations = submodule.findIllegalDependencies(filter(submodules, submodule));
			if (!violations.isEmpty())
				allViolations.put(submodule, violations);
		}

		return new SubmodulesBlueprintCheckResult(submodulesBlueprint, allViolations);
	}

	private static Set<Submodule> filter(List<Submodule> submodules, Submodule submodule) {
		Set<Submodule> submodulesCopy = Sets.newHashSet(submodules);
		submodulesCopy.remove(submodule);
		return submodulesCopy;
	}

	public boolean isWanted(DesignQualityExpectations designQualityExpectations) {
		return designQualityExpectations.getSubmodulesBlueprint() != null;
	}

}

File: src/main/java/org/tindalos/principle/domain/detector/submodulesblueprint/SubmodulesBlueprintViolationDetector.scala
package org.tindalos.principle.domain.detector.submodulesblueprint

import org.tindalos.principle.domain.core.Package
import org.tindalos.principle.domain.coredetector.{AnalysisInput, Detector}
import org.tindalos.principle.domain.expectations.Expectations

import scala.collection.immutable.Map

object SubmodulesBlueprintViolationDetector {
  def buildInstance(buildSubmodules: (String, List[Package], String) => Set[Submodule]) = new Detector {
    override def isWanted(designQualityExpectations: Expectations) = designQualityExpectations.submodulesBlueprint != null

    override def analyze(checkInput: AnalysisInput) = {

      val submodulesBlueprint = checkInput.submodulesBlueprint()

      try {
        val submodules = buildSubmodules(
          submodulesBlueprint.location,
          checkInput.packages, checkInput.analysisPlan.basePackage)

        val (aID, aMD) = problematicDependencies(submodules)

        new SubmodulesBlueprintAnalysisResult(submodulesBlueprint, aID, aMD)
      } catch {
        case ex: OverlappingSubmoduleDefinitionsException => new SubmodulesBlueprintAnalysisResult(submodulesBlueprint, overlaps = ex.overlaps)
      }
    }

    private def problematicDependencies(submodules: Set[Submodule]): (Map[Submodule, Set[Submodule]], Map[Submodule, Set[Submodule]]) = {
      val emptyMap = Map[Submodule, Set[Submodule]]()
      submodules.foldLeft((emptyMap, emptyMap))((acc, submodule) => {
        val otherSubmodules = submodules.filterNot(_.equals(submodule))
        val illegalDependencies = submodule.findIllegalDependencies(otherSubmodules)
        val missingDependencies = submodule.findMissingPredefinedDependencies(otherSubmodules)

        val aID2 = if (illegalDependencies.isEmpty) acc._1 else acc._1 + (submodule -> illegalDependencies)
        val aMD2 = if (missingDependencies.isEmpty) acc._2 else acc._2 + (submodule -> missingDependencies)

        (aID2, aMD2)
      })
    }
  }
}